lib/runtime/wasmer/exports.go

```diff
@@ -347,6 +347,85 @@ func (in *Instance) GrandpaSubmitReportEquivocationUnsignedExtrinsic(
 	return nil
 }
 
+func fixPersistedValidationDataEncodedBytes(bytes []byte) ([]byte, error) {
+	// Trim first byte (01) in the beginning and bytes for RelayParentNumber(4),
+	// RelayParentStorageRoot (32), MaxPovSize(4) to get HeadData bytes
+
+	bytes = bytes[1:]
+	bytesExceptHeadData := bytes[len(bytes)-40:]
+	headDataBytes := bytes[:len(bytes)-40]
+
+	newEncodedHeadData, err := scale.Marshal(headDataBytes)
+	if err != nil {
+		return []byte{}, fmt.Errorf("scale marshalling: %w", err)
+	}
+
+	return append(newEncodedHeadData, bytesExceptHeadData...), nil
+}
+
+// ParachainHostPersistedValidationData returns persisted validation data for the given parachain id.
+func (in *Instance) ParachainHostPersistedValidationData(
+	parachaidID uint32,
+	assumption parachaintypes.OccupiedCoreAssumption,
+) (*parachaintypes.PersistedValidationData, error) {
+	buffer := bytes.NewBuffer(nil)
+	encoder := scale.NewEncoder(buffer)
+	err := encoder.Encode(parachaidID)
+	if err != nil {
+		return nil, fmt.Errorf("encoding equivocation proof: %w", err)
+	}
+	err = encoder.Encode(assumption)
+	if err != nil {
+		return nil, fmt.Errorf("encoding key ownership proof: %w", err)
+	}
+
+	encodedPersistedValidationData, err := in.Exec(runtime.ParachainHostPersistedValidationData, buffer.Bytes())
+	if err != nil {
+		return nil, err
+	}
+
+	fixedEncodedBytes, err := fixPersistedValidationDataEncodedBytes(encodedPersistedValidationData)
+	if err != nil {
+		return nil, fmt.Errorf("fixing persisted validation data bytes: %w", err)
+	}
+
+	persistedValidationData := &parachaintypes.PersistedValidationData{}
+	err = scale.Unmarshal(fixedEncodedBytes, persistedValidationData)
+	if err != nil {
+		return nil, fmt.Errorf("scale decoding: %w", err)
+	}
+
+	return persistedValidationData, nil
+}
+
+// ParachainHostValidationCode returns validation code for the given parachain id.
+func (in *Instance) ParachainHostValidationCode(parachaidID uint32, assumption parachaintypes.OccupiedCoreAssumption,
+) (*parachaintypes.ValidationCode, error) {
+	buffer := bytes.NewBuffer(nil)
+	encoder := scale.NewEncoder(buffer)
+	err := encoder.Encode(parachaidID)
+	if err != nil {
+		return nil, fmt.Errorf("encoding parachain id: %w", err)
+	}
+	err = encoder.Encode(assumption)
+	if err != nil {
+		return nil, fmt.Errorf("encoding occupied core assumption: %w", err)
+	}
+
+	encodedValidationCode, err := in.Exec(runtime.ParachainHostValidationCode, buffer.Bytes())
+	if err != nil {
+		return nil, err
+	}
+
+	var validationCode []byte
+	err = scale.Unmarshal(encodedValidationCode, &validationCode)
+	if err != nil {
+		return nil, fmt.Errorf("scale decoding: %w", err)
+	}
+
+	return (*parachaintypes.ValidationCode)(&validationCode), nil
```

Comment:
I think we can directly unmarshal into `ValidationCode` and avoid using `[]byte`. I was just playing around with it and had committed this code.